File: calims2-webapp/src/web/static/javascript/namingconvention/NameValidator.js

var nameValidator = {
    // Pattern map for the various character sets specified in word
	// Tis needs to be updated in synch with the CharacterSet enumeration	
    patterns : { "ALL_LOWER_CASE" : "[a-z]*$" , "ALL_UPPER_CASE" : "[A-Z]*$" , "MIXED_CASE" : "[a-zA-Z]*$", "DIGITS" : "[0-9]*$"},
    
    // Gets the naming convention that applies to the laboratory with the given id.
    // namingConventionMap The map of naming convention objects
    // laboratoryId The id of the laboratory
    getNamingConvention : function(namingConventionMap, laboratoryId) {
      if (laboratoryId) {
        var result = namingConventionMap[laboratoryId];
        if (result) {
          return result;
        } else {
          return namingConventionMap["global"];
        }  
      } else {
        return namingConventionMap["global"];
      }
    },
    
    // test all the possible violations for the given input (min length, max length and character set)
    // input The input to test (the substring corresponding to the word)
    // word the word to test
    // return true if any violation is found
    testAll : function(input, word) {
      return this.testMinLength(input, word) || this.testMaxLength(input, word) || this.testCharset(input, word);
    },
    
    // test the character set violation
    // input The input to test (the substring corresponding to the word)
    // word the word to test
    // return true if a character set violation is found
    testCharset : function(input, word) {
      var pattern = "^";
      if (word.padCharacter && word.padCharacter.length > 0) {
        pattern += word.padCharacter + "*";
      }
      pattern += this.patterns[word.characterSet];
      var regexp = new RegExp(pattern);
      return !regexp.test(input);
    },
    
    // test the minimum length set violation
    // input The input to test (the substring corresponding to the word)
    // word the word to test
    // return true if a minimum length set violation is found
    testMinLength : function(input, word) {
      return word.minLength && input.length < word.minLength;
    },
    
    // test the maximum length set violation
    // input The input to test (the substring corresponding to the word)
    // word the word to test
    // return true if a maximum length set violation is found
    testMaxLength : function(input, word) {
      return word.maxLength && input.length > word.maxLength;
    },
    
    // Validates an input using the given naming convention.
    // namingConvention The naming convention to use
    // input the input to validate
    validate : function(namingConvention, input) {
      if (!input || (namingConvention.maxLength && input.length > namingConvention.maxLength)) {
        return false;
      }
      var words = namingConvention.words; 
      var toValidate = input;
      for (var i = 0; i < words.length; i++) {
        var word = words[i];
        var wordToValidate = null;
        if (word.delimiter) {
          var pos = toValidate.indexOf(word.delimiter);
          if (pos < 0) {
            return false;
          }
          wordToValidate = toValidate.substring(0, pos);
          toValidate = toValidate.substring(pos + word.delimiter.length);
        } else {
          wordToValidate = toValidate;
          toValidate = "";
        }
        if (this.testAll(wordToValidate, word)) {
          return false;
        }
      }
      return true;
    },
    // Validates an input using the given naming convention map.
    // namingConventionMap The map of naming convention objects
    // input the input to validate
    // laboratoryId the laboratory id
    validateInput : function(namingConventionMap, input, laboratoryId) {
      var result = true;
      if (laboratoryId && laboratoryId != "-1") {
        var namingConvention = this.getNamingConvention(namingConventionMap, laboratoryId);
        if (namingConvention) {
          result = this.validate(namingConvention, input);
        }
      } else {
        result = false;
      }
      return result;
    },  
    // Validates an input using the given naming convention map.
    // namingConventionMap The map of naming convention objects
    // inputId the id of the dijit widget containing the input to validate
    // labId the id of the dijit widget containing containing the laboratory id
    validateName : function(namingConventionMap, inputId, labId) {
      var input = dijit.byId(inputId).getValue();
      var laboratoryId = dijit.byId(labId).getValue();
      return this.validateInput(namingConventionMap, input, laboratoryId);
    }
  };  

File: calims2-api/src/java/gov/nih/nci/calims2/business/administration/namingconvention/NameValidator.java
/*L
 * Copyright Moxie Informatics.
 *
 * Distributed under the OSI-approved BSD 3-Clause License.
 * See http://ncip.github.com/calims/LICENSE.txt for details.
 */

/**
 *
 */
package gov.nih.nci.calims2.business.administration.namingconvention;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.apache.commons.lang.StringUtils;

import gov.nih.nci.calims2.dao.GenericDao;
import gov.nih.nci.calims2.domain.administration.Laboratory;
import gov.nih.nci.calims2.domain.administration.NamingConvention;
import gov.nih.nci.calims2.domain.administration.Word;
import gov.nih.nci.calims2.domain.administration.enumeration.CharacterSet;
import gov.nih.nci.calims2.domain.administration.enumeration.NamingConventionType;

/**
 * Validates a given String given The String to validate and the Naming convention.
 * 
 * @author viseem
 * 
 */
public class NameValidator {
  
  private static final Map<CharacterSet, String> PATTERNS = patternMap();

  private GenericDao genericDao;

  /**
   * Builds the map of patterns per Character Set.
   * 
   * @return the map of patterns per Character Set.
   */
  private static Map<CharacterSet, String> patternMap() {
    Map<CharacterSet, String> patterns = new HashMap<CharacterSet, String>();
    patterns.put(CharacterSet.ALL_LOWER_CASE, "\\p{Lower}");
    patterns.put(CharacterSet.ALL_UPPER_CASE, "\\p{Upper}");
    patterns.put(CharacterSet.MIXED_CASE, "\\p{Alpha}");
    patterns.put(CharacterSet.DIGITS, "\\p{Digit}");
    return patterns;
  }

  /**
   * Gets the naming convention applicable to a given laboratory and entity type.
   * 
   * @param laboratoryId The id of the laboratory
   * @param type The type of entity
   * @return The applicable naming convention or null if none is found.
   */
  public NamingConvention getNamingConvention(Long laboratoryId, NamingConventionType type) {
    String queryName = NamingConvention.class.getName() + ".findByLaboratoryIdAndType";
    Map<String, Object> params = new HashMap<String, Object>();
    params.put("laboratoryId", laboratoryId);
    params.put("type", type);
    List<NamingConvention> namingConventions = genericDao.findByNamedQuery(queryName, params);
    if (namingConventions.size() == 1) {
      return namingConventions.get(0);
    }
    NamingConvention allLabConvention = null;
    for (NamingConvention nc : namingConventions) {
      Laboratory laboratory = nc.getLaboratory();
      if (laboratory != null && laboratory.getId().equals(laboratoryId)) {
        return nc;
      }
      if (laboratory == null) {
        allLabConvention = nc;
      }
    }
    return allLabConvention;
  }

  /**
   * Test if the given string follows the given naming convention.
   * 
   * @param input The input string to test
   * @param namingConvention The naming convention
   * @return true if the input string is valid according to the naming convention
   */
  public boolean validate(String input, NamingConvention namingConvention) {
    if (input == null || (namingConvention.getMaximumLength() != null && input.length() > namingConvention.getMaximumLength())) {
      return false;
    }
    String toValidate = input;
    for (Word word : getWords(namingConvention)) {
      String wordToValidate = null;
      if (StringUtils.isNotEmpty(word.getDelimiter())) {
        int pos = toValidate.indexOf(word.getDelimiter());
        if (pos < 0) {
          return false;
        }
        wordToValidate = toValidate.substring(0, pos);
        toValidate = toValidate.substring(pos + word.getDelimiter().length());
      } else {
        wordToValidate = toValidate;
        toValidate = "";
      }
      if (testAll(wordToValidate, word)) {
        return false;
      }
    }
    return true;
  }

  private boolean testAll(String input, Word word) {
    return testMinLength(input, word) || testMaxLength(input, word) || testCharset(input, word);
  }

  private boolean testCharset(String input, Word word) {
    String pattern = "^";
    if (StringUtils.isNotEmpty(word.getPadCharacter())) {
      pattern += word.getPadCharacter() + "*";
    }
    pattern += PATTERNS.get(word.getCharacterSet()) + "*$";
    boolean result = !input.matches(pattern);
    return result;
  }

  private boolean testMinLength(String input, Word word) {
    return word.getMinimumLength() != null && input.length() < word.getMinimumLength();
  }

  private boolean testMaxLength(String input, Word word) {
    return word.getMaximumLength() != null && input.length() > word.getMaximumLength();
  }

  private List<Word> getWords(NamingConvention namingConvention) {
    List<Word> words = new ArrayList<Word>();
    words.addAll(namingConvention.getWordCollection());
    Collections.sort(words, new WordComparator());
    return words;
  }

  /**
   * @param genericDao the genericDao to set
   */
  public void setGenericDao(GenericDao genericDao) {
    this.genericDao = genericDao;
  }

}
